### PR TITLE
Implement ZIP 32

### PR DESCRIPTION
**Pull request by daira**
Includes miscellaneous improvements, including an additional ``-t json`` output option that does not reverse 32-byte test vectors.

### CODE REVIEW

**Comment by daira** [comment]
Note that GitHub is showing commits out-of-order.

File: sapling_key_components.py [hunk 1]
```diff
@@ -75,10 +61,29 @@ def ak(self):
...
         return PROVING_KEY_BASE * self.nsk()
 
+
+class DerivedIvk(object):
     @cached
     def ivk(self):
         return Fr(crh_ivk(bytes(self.ak()), bytes(self.nk())))
 
+
+class SpendingKey(DerivedAkNk, DerivedIvk):
```
[the remark this answers is not in the document]
Comment (daira) [reply sapling_key_components.py]: These mixins have no overlapping methods, so it doesn't matter.

**Comment by daira** [comment]
Comments addressed.

File: sapling_utils.py [hunk 2]
```diff
@@ -10,26 +10,58 @@ def i2lebsp(l, x):
...
 def leos2bsp(buf):
     return sum([[(c >> i) & 1 for i in range(8)] for c in buf], [])
 
+def bebs2osp(bits, m=None):
+    l = len(bits)
+    bits = [0] * (8 * cldiv(l, 8) - l) + bits
+    return bytes([bebs2ip(bits[i:i + 8]) for i in range(0, len(bits), 8)])
+
+def beos2bsp(buf):
+    return sum([[(c >> (7-i)) & 1 for i in range(8)] for c in buf], [])
```
[the remark this answers is not in the document]
Comment (str4d) [reply sapling_utils.py]: Hmm, this is not used anywhere. Either remove it, or add a test that `beos2ip(x) == bebs2ip(beos2bsp(x))` (and likewise for `leos2ip`, as that is missing).

File: ff1.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    for i in range(9, -1, -1):
+        Q = tweak + b'\0'*((-t-b-1) % 16) + bytes([i]) + bebs2osp(A)
+        y = beos2ip(aes_cbcmac(key, P + Q)[:d])
+        c = (NUM_2(B)-y) % (1<<u)
+        C = STR_2(u, c)
+        B = A
+        A = C
+    return A + B
+
+def test_ff1():
```
Comment (str4d) [inline ff1.py]: Add the two test cases from my Rust implementation in here: https://github.com/str4d/fpe/blob/master/src/ff1.rs#L720-L758

**Pull request merged**